Follow up to "🚸 Update ProUI Plot graph (#26539)"

classicrocker883 opened the pull request:
<!--

Submitting a Pull Request

- Please fill out all sections of this form. You can delete the helpful comments.
- Pull Requests without clear information will take longer and may even be rejected.
- We get a high volume of submissions so please be patient during review.

-->

### Description
I wanted to make an update following #26539

- update ProUI for additional `ICON_`'s for new `DWIN_SET`
- update temp popup
- remove unused,
- rearrange/change `ICON_` names
- rearrange enum PauseMessage
- add TJC support

Update
---
- update DWIN_ICO.py
- add extra `ICON_`'s to **buildroot/share/dwin**
- rename `DACAI_ICON_Show` => `dacaiIconShow` | `ICON_Show` => `iconShow`

I made a fix for this issue "[BUG] Unresponsive display with Hotend/Bed Temp Graph - PROUI_ITEM_PLOT https://github.com/MarlinFirmware/Marlin/issues/26630 "

- Update `ID_MPC/PIDProcess` with `TERN_`
- Fix Plot Temp Graph issue by making it return screen if there is a popup

<br>
<!--

Clearly describe the submitted changes with lots of details. Include images where helpful. Initial reviewers may not be familiar with the subject, so be as thorough as possible. You can use MarkDown syntax to improve readability with bullet lists, code blocks, and so on. PREVIEW and fix up formatting before submitting.

-->

### Requirements
ProUI - `DWIN_LCD_PROUI`
<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits
this updates **proui/dwin.cpp** menu ICON's with more variety.

- fixes some things that are unused and duplicates
- tweak it to be arranged better

<!-- What does this PR fix or improve? -->

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues
#### for the added `EXTUI`, I made some fixes to the errors which popped up.
- reverted `allow_cold_extrude_override` in **src/module/temperature.cpp** (which was removed??)  
- fixed the `no such file` errors in moved **lcd/dogm/u8g/**
- various small tweaks to fix errors/warnings since `EXTUI` stuff was added.
<!-- Does this PR fix a bug or fulfill a Feature Request? Link related Issues here. -->

classicrocker883 committed:
update DWIN for additional ICON_'s
update temp popup
remove unused,
rearrange/change ICON_ names
rearrange enum PauseMessage
add TJC support

classicrocker883 committed:
update dwin_defines.h, add TJC option

thinkyhead committed:
remove extra method impl

thinkyhead committed:
misc. cleanup

thinkyhead committed:
misc. style

thinkyhead committed:
fix duplicate icons

thinkyhead committed:
caps

thinkyhead committed:
ICON_ResumeEEPROM => ICON_ResetEEPROM

thinkyhead commented:
You'll need to rename Icon number 44 in the various icon packs from `ICON_ResumeEEPROM` to `ICON_ResetEEPROM`. I renamed this for clarity since it's being used as a stand-in for a more general "reset" icon.

thinkyhead commented:
For best compatibility with Marlin's DWIN helper scripts, you'll want to organize custom icon packs into "`icons-#`" sub-folders with the same numbering and organization as the `buildroot/share/dwin` folder contents.

thinkyhead commented:
What is the procedure to prepare icons 93, 200-206, and 250-254 ? and do these only pertain to TJC? We should add these to `buildroot/share/dwin` for completeness with our "unified" DWIN set.

thinkyhead commented:
Ah, I see those icons are only pertinent to `DWIN_CREALITY_LCD_CUSTOM_ICONS` for JyersUI. Since that UI is no longer being maintained and has fallen far behind ProUI we may just remove it from Marlin altogether. Efforts are best spent on improving ProUI. It's doubtful that ProUI will ever support languages other than those that use the Roman alphabet., so we've retained CrealityUI because it supports Chinese language.

Another thing we can do to improve support for various UI offerings is to add an ExtUI layer to each of them and add any new event handlers that ExtUI doesn't already have. This would eliminate the need to have so many custom UI callbacks scattered throughout Marlin to accommodate particular UIs.

thinkyhead committed:
not all sets

thinkyhead committed:
updated icons

thinkyhead committed:
add missing icons (prob. need optimization)

thinkyhead committed:
update icon folders

thinkyhead commented:
I've updated the contents of the `buildroot/share/dwin` folder, so it should now have everything needed to generate the numbered ICO files for sets 3, 4, and 7. If there's anything missing from that folder, now would be the best time to add the missing pieces.

classicrocker883 commented:
> You'll need to rename Icon number 44 in the various icon packs from `ICON_ResumeEEPROM` to `ICON_ResetEEPROM`.

I suppose you mean within this **PR**  [here, Configurations #994](https://github.com/MarlinFirmware/Configurations/pull/994)?  
Redoing each one is tedious just to rename one ICON name. I suppose I could do it, but it really doesn't matter with firmware implementation. I would only be able to do it for any labeled `DWIN_SET` and not for the other _TJC_ "private" or _DACAI_.

let me know if you want me to do that. otherwise how it is now since you renamed  to `ICON_ResetEEPROM` will be fine.

classicrocker883 commented:
> What is the procedure to prepare icons 93, 200-206, and 250-254 ? and do these only pertain to TJC? We should add these to `buildroot/share/dwin` for completeness with our "unified" DWIN set.  
  
_*it is actually 249-254_

I do not know what others are supported in TJC, same goes for the "private" folder. I guess the same original icons work, but the added ones 93, 200-206, and 249-254 probably do not...

What do you mean by to prepare the icons?

so what I did was take the existing ._jpg_ files, edit them in MS paint, literally pixel by pixel (since it is only 20x20 for the majority of them); remove the "noise" from the background, clean up anything to match the others, fix the loss of quality by the previous editor which happens when saving as ._jpg_. then I saved it as ._png_, then used something to convert to ._jpg_ without losing any quality. after going through each one, I followed the instructions how to convert all of them to an ._ICO_ and thats how you get your `DWIN_SET`.

classicrocker883 commented on Marlin/src/lcd/e3v2/proui/dwin.cpp:
```diff
@@ -482,8 +471,7 @@ void dwinResetStatusLine() {
   uint32_t hash = 5381;
-  char c;
-  while ((c = *str++)) hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
+  for (char c; (c = *str++);) hash = ((hash << 5) + hash) + c; /* hash * 33 + c */
```
I just noticed that `getHash` doesn't seem to be used anywhere. I can't find any other references for it.

could then it be removed?? or whats it for?

thinkyhead replied:
Hashing a string allows the detection of a string change without storing a copy of the old string. We check for a string change to avoid unnecessary expensive drawing operations.

classicrocker883 committed:
add extra DWIN_SET's as icons-9

classicrocker883 commented:
> I've updated the contents of the `buildroot/share/dwin` folder, so it should now have everything needed to generate the numbered ICO files for sets 3, 4, and 7. If there's anything missing from that folder, now would be the best time to add the missing pieces.

I added the icon packs of the other `DWIN_SET`'s as "icons-9". is this what you were looking for?

classicrocker883 commented on Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp:
```diff
@@ -232,46 +233,50 @@ bool BedLevelTools::meshValidate() {
-    }
+      else {          // has value
+        MString<12> msg;
+        constexpr bool is_wide = (GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10);
+        if (is_wide)
+          msg.setf(F("%02i"), uint16_t(z * 100) % 100);
+        else
+          msg.set(p_float_t(abs(z), 2));
```
@thinkyhead was there a mistake made here?

originally it was like this
```y
    if ((GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10))
      msg.set(p_float_t(abs(z), 2));
    else
      msg.setf(F("%02i"), uint16_t(abs(z - int16_t(z)) * 100));
```
if  
`bool is_wide = (GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10);` 
 
 if that is `TRUE` as the above, wouldnt it remain the same like:  
```y
    if (is_wide)
      msg.set(p_float_t(abs(z), 2));
    else
      msg.setf(F("%02i"), uint16_t(z * 100) % 100);
```


also, when you changed `msg.setf` to get the decimal, I was wondering if that were more efficient. this is the answer I got:

>uint16_t(z * 100) % 100:
>
>This expression multiplies z by 100, converts the result to an unsigned 16-bit integer (uint16_t), and then takes the modulo 100. This approach might be less efficient because it involves a floating-point multiplication and conversion to an integer.

so with that, I'm wondering if avoiding floating-point multiplication is better?

then it says
>However, the actual performance difference might be minimal, and modern compilers are often able to optimize such expressions.

classicrocker883 committed:
update bedlevel_tools

thinkyhead committed:
oops, meant ">="

thinkyhead committed:
Not hosting old icon sets, only our unified set

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26563

thinkyhead commented:
> I added the icon packs of the other DWIN_SET's as "icons-9". is this what you were looking for?

Not excatly. We're not supporting the original Creality set (icon set 9), only Marlin's (icon sets 3, 4, 7). So I'm just checking to make sure we have all the icons needed for that. It looks like ProUI is now using Icons numbered 93, 200-206, 250-254, so we need the standard icons added to Icon Set 7. I presume that Miguel has already added those icons and created new `DWIN_SET` / `private` folders that contain the updated icons for his distribution. We'll need those icons added to our own set 7 in order to support this updated ProUI code.

Note that after we get through this set of changes you should submit any changes we've made to mainline Marlin to the @mriscoc fork so that we are relatively in sync. In the long run it will be better for you to first submit any changes and fixes to that fork, and then we can coordinate with Miguel to bring those changes over _en masse_ to this repo when he reaches a good milestone. The same goes for any other forks that are actively under development.

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
small change - add spacing dwinui.h

classicrocker883 committed:
update checkkey ID_xx
arrange
remove unused ID_Reboot
add MPC_PROCESS
use TERN_ for ID_PIDProcess

classicrocker883 committed:
remove unused, add DACAI support (maybe not full support??)

classicrocker883 committed:
fix error - dwinDrawBox undef

classicrocker883 committed:
move MSG_PREHEAT_(count)

classicrocker883 commented:
question:  
is there a way to make this work for all languages?  
besides placing this in every **language_xx.h** file, since they are `using namespace Language_en;` anyway...  
take the following in **proui/dwin.h** and place it in **language/language_en.h**
```y
  #define _MSG_PREHEAT(N) \
    LSTR MSG_PREHEAT_##N                  = _UxGT("Preheat ") PREHEAT_## N ##_LABEL; \
    LSTR MSG_PREHEAT_## N ##_SETTINGS     = _UxGT("Preheat ") PREHEAT_## N ##_LABEL _UxGT(" Conf");
  #if PREHEAT_COUNT > 1
    REPEAT_S(2, INCREMENT(PREHEAT_COUNT), _MSG_PREHEAT)
  #endif
```
<br>

so then for instance, using the following as example  
 `LSTR MSG_PREHEAT_M            = _UxGT("Preheat $");`  
 `LSTR MSG_PREHEAT_M_SETTINGS   = _UxGT("Preheat $ Conf");`

could we just use it like this?  
```y
  #define _MSG_PREHEAT(N) \
    LSTR MSG_PREHEAT_##N                  = MSG_PREHEAT_M PREHEAT_## N ##_LABEL; \
    LSTR MSG_PREHEAT_## N ##_SETTINGS     = MSG_PREHEAT_M_SETTINGS PREHEAT_## N ##_LABEL;
  ...
```
would this just replace the `$`? or how would that work?

thinkyhead committed:
Not the way to do this

thinkyhead commented on Marlin/src/lcd/language/language_en.h:
```diff
@@ -138,6 +138,13 @@ namespace LanguageNarrow_en {
+  #if PREHEAT_COUNT > 1
+    REPEAT_S(2, INCREMENT(PREHEAT_COUNT), _MSG_PREHEAT)
+  #endif
+
```
> so then for instance, using the following as example

This method of producing strings for DWIN LCD is obsolete, so we shouldn't propagate it to the language files. Instead we do substitution using a string like "Preheat $" where the `$` is replaced by a string supplied in the string emitting function. This block is still retained in `dwin.h` for the benefit of ProUI but it should be getting phased out by @mriscoc at some point soon, if not already.

thinkyhead committed:
add new icons to script

thinkyhead commented:
> would this just replace the $? or how would that work?

I didn't pay closest attention when Miguel implemented the string substitution stuff in the last update, and after a cursory look at the menu code in `dwin.cpp` it looks like basic language support with string substitution is still incomplete in ProUI. Hopefully that will be added pretty soon. In the meantime we'll just retain the current workaround.

thinkyhead committed:
remove non-translated atoms

classicrocker883 committed:
quick spacing - dwin_api.h

classicrocker883 committed:
Merge branch 'bugfix-2.1.x-Dec3' of https://github.com/classicrocker883/Marlin into bugfix-2.1.x-Dec3

classicrocker883 commented:
you know the change in "[remove non-translated atoms](https://github.com/MarlinFirmware/Marlin/pull/26563/commits/620aa67b4ff2b31eeb44014b4ec56f227a924f2a)", commit [620aa67](https://github.com/MarlinFirmware/Marlin/pull/26563/commits/620aa67b4ff2b31eeb44014b4ec56f227a924f2a)  
how about making that a language string?:
```y
  LSTR MSG_PID_SET_KP   = _UxGT("Set" STR_KP);
  LSTR MSG_PID_SET_KI   = _UxGT("Set" STR_KI);
  LSTR MSG_PID_SET_KD   = _UxGT("Set" STR_KD);
```

thinkyhead committed:
fix script

classicrocker883 commented:
> it looks like basic language support with string substitution is still incomplete in ProUI. Hopefully that will be added pretty soon. In the meantime we'll just retain the current workaround.

is that what you mean? to basically replace all the strings as `MSG_<whatever>`?

I know in his [ProUI fork](https://github.com/mriscoc/Ender3V2S1), almost all the menu strings seem to be incorporated within **language_en.h** namespace  
for quick reference, at near the end of it:
```y
  // DWIN_LCD_PROUI specific message strings
  LSTR MSG_BED_TEMPERATURE      = _UxGT("Bed Temperature");
  LSTR MSG_NOZZLE_TEMPERATURE   = _UxGT("Nozzle temperature");
  LSTR MSG_TOO_HIGH             = _UxGT("is too high");
  LSTR MSG_TOO_LOW              = _UxGT("is too low");
  LSTR MSG_HOTEND_PID_SETTINGS  = _UxGT(STR_HOTEND_PID " Settings");
  LSTR MSG_BED_PID_SETTINGS     = _UxGT(STR_BED_PID " Settings");
```

but I was thinking... instead of placing that at the end of the main `LanguageNarrow_en` namespace  
we can use the next namespace since it uses `HAS_DWIN_E3V2`.
```y
namespace LanguageWide_en {
  using namespace LanguageNarrow_en;
  #if LCD_WIDTH >= 20 || HAS_DWIN_E3V2
```

thinkyhead committed:
Fix splitIco.py

thinkyhead commented:
> I know in his [ProUI fork](https://github.com/mriscoc/Ender3V2S1), almost all the menu strings seem to be incorporated within language_en.h namespace. For quick reference, at near the end of it…

It's fine for the "narrow" language strings to be there if they are 20 characters or less. But the wider strings should definitely be moved to the `LanguageWide_en` namespace.

> is that what you mean? to basically replace all the strings as `MSG_<whatever>`?

All strings that appear on an LCD should use the defined `MSG_` strings and `GET_TEXT` / `GET_TEXT_F` macros to access them. These transparently manage language translation.

But in terms of "substitution" any string with a special character (`$*@{`) filtered through the appropriate lcd string methods will get substitutions regardless of the string's source.

thinkyhead commented:
Our `Configurations` repository is the official home for the Marlin 2.1.x-compatible DWIN `DWIN_SET` folder, DACAI `private` folder, and (now) the `tjc.tft` set. So all those sets need to get the new icons added. I already went ahead and added the new icons to `DWIN_SET/7.ICO`. So they just need to be added to the DACAI `private` set and `tjc.tft`. At that point this PR will be compatible with the configs and both can be merged.

I can go ahead and put the new JPEG files into the `private` folder, but I'm not versed on how to make the `7.zico` file. Maybe the format is obvious after comparing its contents to the size and length of the JPGs. We'll see….

mriscoc commented:
> I presume that Miguel has already added those icons and created new `DWIN_SET` / `private` folders that contain the updated icons for his distribution.

No, I use the default Creality Icon set and not added by default any custom icon to try to avoid screen firmware updates, but I added a custom function to allow to use and extended icon library "10.ICO" (which can fallback to "9.ICO") if custom icons are needed.

Currently, Marlin ProUI has diverged a lot from my original implementation and it is a little difficult to get them in sync again.

thinkyhead commented:
> I use the default Creality Icon set and not added by default any custom icon to try to avoid screen firmware updates

Hmm, well on the one hand, I don't want users to lose the ability to use ProUI without a display update. On the other hand, now we have this PR here with improved cleaner icons, so it makes sense that your Professional Firmware should be able to take advantage of these improvements as well. And really, if you were to insist that your users must update the screen to use the next version of ProUI, I believe very few would complain. It's a pretty easy process and users will consider the improvements over CrealityUI worth the extra step.

> Currently, Marlin ProUI has diverged a lot from my original implementation and it is a little difficult to get them in sync again.

I'm here to help if you run into snags. You can get in touch with me pretty quickly on Discord whenever a question arises.

We're always aiming for better standardization. Consistent language translations and menu titles is one area that we would really like to get nailed down. I also plan to add an ExtUI layer to ProUI and other displays so they no longer require custom callbacks to be scattered throughout the code but will just intercept a single ExtUI event callback at each location. Where ExtUI doesn't currently have an event callback that ProUI or another display needs, we will simply extend the ExtUI API.

If you plan to keep developing ProUI as we enter the Next Phase then we'll need to work more closely to develop sensible APIs and utilities that fit your needs and the needs of similar UIs. It will be important to keep your fork as close to the main fork to keep in sync without a lot of duplicated effort.

A problem with divergent forks of Marlin is that other contributors end up cherry-picking bits and pieces and bringing them here for integration. Since there's no compelling reason for us to refuse a good improvement, we obediently fix it up and integrate it, but it can never match up to the completeness of your own work. So it's always better if your changes are in sync with ours, and vice-versa. That is what we ought to strive for on an ideal planet.

This PR has consumed a few days of productive time. I would much rather be working directly with you instead of intermediaries on these APIs and integrations. (The author of this PR is cool, but we get some real "characters" sometimes.)

> I am not convinced that implementing the substitution routines in ProUI will give a more compact compiled code

We can still do a lot of things to optimize the code and keep it from exploding, so don't hesitate to proceed on that concern. The "LCD String API" is another important area of standardization. You'll be needing it if you want ProUI to use translated strings that have different orderings, adapt to changes in the number of extruders and preheat presets, and so on. It will not increase code size because we can do things like replace 5 menu items with a loop that only emits one menu item 5 times, but with different E indexes (for example). And only a small number of menu items will actually be affected. The string interpolation function is also quite small.

> I'm struggling for every byte due to the program memory limitations in STM32F4XXX RCT6 SoC's

Welcome to my world. I still lose sleep worrying about old Melzi board support. My background is in tiny 6502 computers where every byte counts. So you couldn't ask for a better partner in space-saving than "yours truly."

Anyway, stay in touch! We'll get these things sorted out in time, I'm sure. Note that I am planning a 2.1.3 (or 2.2.0??) release pretty soon, now that Fixed-Time Motion is integrated and we have a long list of bug-fixes. After that release I will be working on LCD standardization in earnest. I'll be sure to let you know as soon as I get started on that process.

classicrocker883 committed:
update ICON_'s - arrange
add inline drawCenteredString

thinkyhead committed:
Use u8g 0.5.2 for now

classicrocker883 commented:
>So they just need to be added to the DACAI private set and tjc.tft.
>I can go ahead and put the new JPEG files into the private folder, but I'm not versed on how to make the 7.zico file. Maybe the format is obvious after comparing its contents to the size and length of the JPGs. We'll see….

that would be great! I couldn't find any information about it myself. I was able to extract non-working .jpg using **Split-ico.py** from the .zico file, that is it was able to grab all the files, but not give anything to work with. maybe someone at Creality or where ever .zico, TJC, DACAI... originates from knows and has the proper tools.

>On the other hand, now we have this PR here with improved cleaner icons, so it makes sense that your Professional Firmware should be able to take advantage of these improvements as well. 

some of the ICONs (anything > #91) already exist in the **DWIN_SET**s which many users may already have installed - especially if they are coming from JyersUI. so why not use them. plus while I was cleaning up the original ones, I added one of my own, **093-ICON_Printer_0**.jpg (generic Ender-3 style). Its good to use for when Homing or an error. I left it open-ended with **Printer_1** for maybe a different type or whatever. unfortunately, I wasnt able to do anything about #78-81 which is 73x66 px. I mean I was able to clean it up, but after installing it appeared to go through a shredder. So I just left them original despite the kinda of pixel haze (typical when saving into .jpg).

furthermore, if anyone wishes to edit .jpg, unless you have a professional app which saves .jpg without losing quality, you have to save as .png. then I used a website which is able to convert .png to .jpg and it doesn't mess with the image. thats how I got it cleaned up, and no doubt why they became messy to begin with.

>I am not convinced that implementing the substitution routines in ProUI will give a more compact compiled code  

I think when I changed a lot of the " _string_ " to `MSG_` in my own, I noticed it reduced the byte size in the end.

thinkyhead committed:
fix at 0.5.4

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26563

classicrocker883 committed:
fix DWIN_ICO.py - was not grabbing all with Split
will now output all files even if UNKNOWN (unlisted)

classicrocker883 committed:
update buildroot/share/dwin/bin README.md

classicrocker883 committed:
add option to center Title

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
change DWIN popup img M0_M1.cpp

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
fix extra space + ` => '

classicrocker883 committed:
more comment fixings

classicrocker883 committed:
revert comment fixings

classicrocker883 committed:
fix for Plot Temp Graph

classicrocker883 committed:
Update Configuration_adv.h

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
fix ReturnScreen while not printing
still need to test

classicrocker883 committed:
fix for last commit, for real this time

classicrocker883 committed:
fix for ID_PlotProcess
needs test

classicrocker883 committed:
fix typo

classicrocker883 committed:
fix typo

classicrocker883 committed:
final fix for Plot graph

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
fix for error

classicrocker883 committed:
fix for indent warning

classicrocker883 committed:
add warning for ICON (.ICO)

classicrocker883 committed:
rename to dacaiIconShow

classicrocker883 committed:
rename ICON_Show

classicrocker883 committed:
move PID case error

classicrocker883 committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x-Dec3

classicrocker883 committed:
flx error, typo

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 commented:
this is already ready for review

thinkyhead commented on Marlin/src/lcd/e3v2/common/dwin_set.h:
```diff
@@ -25,7 +25,10 @@
+#define ICON 7 // Stock icons = 7.ICO; Custom icon DWIN_SET may be different
+#ifdef ICON
+  #warning "Check your DWIN_SET .ICO file for this value of ICON"
+#endif
```
This is asking a lot of users. Our instructions tell users to use our provided `DWIN_SET`. I'm not sure I would even know how to respond to this message during a Marlin build.

classicrocker883 replied:
I suppose that can be reverted. The thing is some `DWIN_SET`s have other than **7.ICO**. I couldn't tell you why the change, basically any other custom display set uses **9.ICO** (which would mean it would need to be `#define ICON 9` instead of "7" if that user has or will install the `DWIN_SET` on their LCD).

It would make sense to have them all equal the same #.ICO since a user is only going to be using one at a time anyway.

I just know from other forums I see users having difficulty regarding this. say they want to install a custom display, the `DWIN_SET` has **9.ICO**, they go to install this firmware with `ICON 7` and the display ICON's dont show.

if that happens they either can change it through the firmware code like we see here, or they rename the .ICO file.

we can probably either change one or the other to make it easier for everyone else:  
make it so that it is...  
1. `#define ICON 9`  
1a. and change the default provided `DWIN_SET` to **9.ICO**  
2. or go into the other custom `DWIN_SET`s and change it to **7.ICO**

Actually, I would simply change it for **9.ICO** and `ICON 9` instead of 7.  
the reason why is most users already have a custom `DWIN_SET` installed on their LCD. so thats why I added a warning when compiling to check/change this value to match the one installed.

maybe this warning isnt called for, but just a comment will do.

thinkyhead replied:
> The thing is some DWIN_SETs have other than 7.ICO

What we want is for users who are running mainline Marlin to use the `DWIN_SET` that we supply for mainline Marlin, which works with all the Ender-3 V2 UI options. We also provided the option in the past to use the Creality UI without updating the display, but that was the only UI giving that option. At this point that might not be an option anymore, but if that option can be retained it will ease migration for any users who just want to update the firmware without modifying their configs or picking a different UI.

thinkyhead replied:
This is what used to be here, which we got rid of (not by mistake):
```cpp
//#define USE_STOCK_DWIN_SET  // Use the Creality stock DWIN_SET instead of Marlin's unified DWIN_SET by The-EG & thinkyhead
#ifdef USE_STOCK_DWIN_SET
  #define ICON 9 // 9.ICO
#else
  #define ICON 7 // 7.ICO
#endif
```

classicrocker883 replied:
how about then:
```y
//#define USE_STOCK_DWIN_SET  // Use the Creality stock or a Custom DWIN_SET instead of Marlin's unified DWIN_SET by The-EG & thinkyhead  
#ifdef USE_STOCK_DWIN_SET  
  #define ICON 9 // 9.ICO  
#else  
  #define ICON 7 // 7.ICO  
  #warning "If you have a Creality or Custom DWIN_SET installed on your display, you may need to enable USE_STOCK_DWIN_SET."  
#endif  
```

thinkyhead replied:
We removed the `USE_STOCK_DWIN_SET` define because the code that supported that set has been completely removed, and users must use our `DWIN_SET` with stock Marlin.

thinkyhead committed:
dwinPidTuning only exists if PROUI_PID_TUNE

thinkyhead committed:
dwinDrawPIDMPCPopup needed for both

thinkyhead commented:
Before this can be merged I want to make sure we have complete sets for DWIN, DACAI, and TJC and that they are all fully compatible with this update. Last I checked we didn't even know how to update the TJC set.

classicrocker883 commented:
> Before this can be merged I want to make sure we have complete sets for DWIN, DACAI, and TJC and that they are all fully compatible with this update. Last I checked we didn't even know how to update the TJC set.

I want to say that this **PR** should be ready to merge on its own. regarding DACAI and TJC - having their Macros defined should enable compatibility.  
I think you are concerned about the updated `DWIN_SET`? Although I have yet to figure out changing/updating a different type of set, but I have already compensated for that in the other **PR** I have.

ex:
```y
dwinShowPopup(TERN(TJC_DISPLAY, ICON_BLTouch, ICON_Printer_0)
```
since they dont have the new `ICON_Printer_0`, it uses one it does have.  
I should also update this so it includes DACAI as well.

classicrocker883 committed:
add DACAI compatibility

classicrocker883 commented on Marlin/src/module/temperature.cpp:
```diff
@@ -724,12 +724,12 @@ volatile bool Temperature::raw_temps_ready = false;
 
     TERN_(EXTENSIBLE_UI, ExtUI::onPidTuning(ExtUI::result_t::PID_STARTED));
-    TERN_(DWIN_LCD_PROUI, dwinPidTuning(isbed ? PIDTEMPBED_START : PIDTEMP_START));
+    TERN_(PROUI_PID_TUNE, dwinPidTuning(isbed ? PIDTEMPBED_START : PIDTEMP_START));
```
@thinkyhead are you sure to revert this?

because the code is encompassed within `HAS_PID_HEATING`.

having `PROUI_PID_TUNE` is a bit redundant, no?

classicrocker883 replied:
if you look at say `TERN_(DWIN_LCD_PROUI, dwinMPCTuning(MPC_TEMP_ERROR));` (line 1147), this is within `MPC_AUTOTUNE`.

im going to remove `PROUI_PID_TUNE` because it simply isn't needed. if you look at **Conditionals_post.h**   
```y
#if HAS_TEMP_CHAMBER && HAS_HEATER_CHAMBER
...
#else
  #undef PIDTEMPCHAMBER
#endif

// PID heating
#if ANY(PIDTEMP, PIDTEMPBED, PIDTEMPCHAMBER)
  #define HAS_PID_HEATING 1
#endif
```

`PIDTEMPCHAMBER` gets undefined anyway, plus there are all kinds of guards when either `PIDTEMP` or `PIDTEMPBED` after

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
remove PROUI_PID_TUNE, replace w/ HAS_PID_HEATING

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

thinkyhead committed:
All E3V2 require Marlin-provided DWIN_SET

thinkyhead committed:
tweak ico script / readme

thinkyhead committed:
ExtUI layer for ProUI

classicrocker883 committed:
Update dwin.h tempcontrol enum

thinkyhead committed:
move lcd/u8g files

thinkyhead commented:
I spent a few days to put more methods into `ExtUI` for some events it wasn't handling yet and made an ExtUI layer for ProUI to replace several of the ProUI-specific calls throughout the code. There are still some ProUI-specific customizations, but most have been shifted over to `ExtUI`. I don't expect the initial attempt to pass CI testing, so pardon the mess as I try to get it patched up without piling on several new commits….

mriscoc commented:
> I spent a few days to put more methods into `ExtUI` for some events it wasn't handling yet and made an ExtUI layer for ProUI to replace several of the ProUI-specific calls throughout the code. There are still some ProUI-specific customizations, but most have been shifted over to `ExtUI`. I don't expect the initial attempt to pass CI testing, so pardon the mess as I try to get it patched up without piling on several new commits….

For referencie some work was already made some time ago for port ProUI to ExtUI here: https://github.com/MarlinFirmware/Marlin/pull/21328 of course that work is now obsolete. I think that moving ProUI to ExtUI should have a separate PR.

classicrocker883 commented on Marlin/src/lcd/e3v2/proui/dwin.cpp:
```diff
@@ -1633,18 +1619,16 @@ void dwinLevelingDone() {
-        #endif
-        default: break;
-      }
+  void dwinStartM303(const int count, const heater_id_t hid, const celsius_t temp) {
```
`hmiData.pidCycles = c;`
should that be `count` instead of `c`?

classicrocker883 commented on Marlin/src/lcd/e3v2/proui/dwin.cpp:
```diff
@@ -1633,18 +1619,16 @@ void dwinLevelingDone() {
-      }
+  void dwinStartM303(const int count, const heater_id_t hid, const celsius_t temp) {
+    hmiData.pidCycles = c;
+    switch (hid) {
```
`hmiData.pidCycles = c;`
should that be `count` instead of `c`?

I think we should keep this as `c` instead of `count`, because **gcode/temp/M303.cpp** has it as `c`

thinkyhead replied:
More descriptive names are better for generic function definitions. It's not uncommon for short-lived G-code internal variables to use single letters to hint they pertain to a parameter letter. Each naming style fits in context and they don't necessarily need to influence each other.

classicrocker883 commented on Marlin/src/gcode/temp/M303.cpp:
```diff
@@ -79,7 +76,7 @@ void GcodeSuite::M303() {
 
-  TERN_(DWIN_LCD_PROUI, dwinStartM303(seenC, c, seenS, hid, temp));
+  TERN_(EXTENSIBLE_UI, ExtUI::onStartM303(c, hid, temp));
 
```
`dwinStartM303()` still used, you mean to covert this over?

also `seenC` and `seenS` go unused, those should be changed in some way??

thinkyhead replied:
I moved the call to `dwinStartM303()` to the implementation of `ExtUI::onStartM303` and it just uses the `count` and `temp` decided above.

thinkyhead committed:
fix typos

thinkyhead committed:
axis_to_axis_t

thinkyhead committed:
ExtUI:: prefix

classicrocker883 committed:
update pid_autotune_menu

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
Merge branch 'bugfix-2.1.x' of https://github.com/MarlinFirmware/Marlin into bugfix-2.1.x-Dec3

classicrocker883 committed:
fixes for errors - revert temperature.cpp 'allow_cold_extrud_override', ui_api.h move 'default: return' to end, rename Pid/Mpc => PID/MPC

classicrocker883 committed:
fix for error: cannot convert

classicrocker883 committed:
fix for not declared

classicrocker883 committed:
multiple changes to change warnings and errors. ui_api.h comment to keep/remove?

classicrocker883 committed:
fix no such file, lcd/dogm/u8g

classicrocker883 committed:
fix no such file, lcd/dogm/u8g pt.2

classicrocker883 committed:
revert '#include "lcdprint.h" - marlinui.h

classicrocker883 committed:
typo in ftdi_eve_extui.cpp - cstr, fstr

classicrocker883 committed:
??fix for abl/G29.cpp??
either it is supposed to be TERN or TERN_, if the latter then ' : LEVELING_BED_TEMP' is redundant.

classicrocker883 committed:
fix no such file lcd/dogm/u8g

classicrocker883 committed:
fix 'HAS_PLR_UI_FLAG', move #define

classicrocker883 committed:
fixes for DWIN and ProUI

thinkyhead committed:
Also ia_creality

thinkyhead committed:
other IA updates

thinkyhead committed:
cleanup

classicrocker883 committed:
update spacing, revert 'stopPrint', change stuff to ExtUI in ProUI

thinkyhead committed:
add chamber pid tuning

thinkyhead committed:
DGUS_MSG_TEMP_TOO_LOW => MSG_TEMP_TOO_LOW

thinkyhead committed:
move timeout msg

thinkyhead committed:
oops

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26563

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26563

thinkyhead committed:
merge followup

classicrocker883 committed:
split result_t to pidresult_t / mpcresult_t

convicte commented:
This seems to have had some unintended consequences breaking LCD support for among others Anycubic Viper

```
Marlin\src\inc\Warnings.cpp:115:6: warning: #warning "Note: Auto-assigned LCD_SERIAL_PORT. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)" [-Wcpp]
  115 |     #warning "Note: Auto-assigned LCD_SERIAL_PORT. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)"
      |      ^~~~~~~
Marlin\src\inc\Warnings.cpp:223:6: warning: #warning "Note: Auto-assigned Z2 STEP/DIR/ENABLE_PINs to unused En_STEP/DIR/ENABLE_PINs. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)" [-Wcpp]
  223 |     #warning "Note: Auto-assigned Z2 STEP/DIR/ENABLE_PINs to unused En_STEP/DIR/ENABLE_PINs. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)"
      |      ^~~~~~~
Marlin\src\inc\Warnings.cpp:760:4: warning: #warning "GD32 based controllers may not be fully compatible with STM32 platforms. Please report any issues." [-Wcpp]
  760 |   #warning "GD32 based controllers may not be fully compatible with STM32 platforms. Please report any issues."
      |    ^~~~~~~
c:/users/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/trigorilla_v006/src/src/gcode/config/M302.cpp.o: in function `GcodeSuite::M302()':
C:\Users\Documents\GitHub\Marlin/Marlin\src\gcode\config/M302.cpp:54: undefined reference to `ExtUI::onSetMinExtrusionTemp(short)'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\trigorilla_v006\firmware.elf] Error 1
```
